### PR TITLE
feat: raw transaction batch submitter

### PR DESCRIPTION
**Pull request by karlfloersch**
**Description**
Re-introduce raw tx based batch submitter first implemented here: https://github.com/ethereum-optimism/batch-submitter/pull/56

**Additional context**
This contributes to https://github.com/ethereum-optimism/optimism/issues/496

### CODE REVIEW

File: packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts [hunk 1]
```diff
@@ -202,9 +202,10 @@ export class StateBatchSubmitter extends BatchSubmitter {
       [...Array(blockRange).keys()],
```
Comment (tynes) [inline packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts:202]: If `blockRange` is negative, an error will be thrown

File: packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts [hunk 2]
```diff
@@ -202,9 +202,10 @@ export class StateBatchSubmitter extends BatchSubmitter {
...
       async (i: number) => {
         this.log.debug('Fetching L2BatchElement', { blockNo: startBlock + i })
-        const block = (await this.l2Provider.getBlockWithTransactions(
-          startBlock + i
-        )) as L2Block
+        const block = (await this.l2Provider.send('eth_getBlockByNumber', [
+          toRpcHexString(startBlock + i),
+          true,
+        ])) as L2Block
```
Comment (tynes) [inline packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts:208]: This particular change is due to migration away from the `OptimismProvider`?

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 3]
```diff
@@ -742,70 +706,41 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
+      true,
+    ])) as L2Block
   }
 
   private _isSequencerTx(block: L2Block): boolean {
-    return block.transactions[0].queueOrigin === QueueOrigin.Sequencer
+    return (
+      queueOriginPlainText[block.transactions[0].queueOrigin] ===
+      QueueOrigin.Sequencer
+    )
```
Comment (gakonst) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:744]: Why is this needed? This should be getting set [here](https://github.com/ethereum-optimism/optimism/blob/master/packages/core-utils/src/l2context.ts#L34-L35) already, no need for plaintext stuff?

**Review by gakonst — CHANGES_REQUESTED** [review]

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 3 above)
Comment (annieke) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:744]: ^i think this may no longer be needed with `injectL2Context`, which wasn't used the last time we made this change to batch submitter

File: packages/batch-submitter/src/utils.ts [hunk 4]
```diff
@@ -21,3 +21,10 @@ export const remove0x = (str: string): string => {
     return str
   }
 }
+
+export const formatNumber = (num: any): number => {
+  if (num === '0x') {
+    return 0
+  }
+  return BigNumber.from(num).toNumber()
+}
```
[the remark this answers is not in the document]
Comment (annieke) [reply packages/batch-submitter/src/utils.ts:30]: i believe i deprecated the entire `utils` file in batch submitter while adding logs, since everything is either duplicated in core-utils or unused: https://github.com/ethereum-optimism/batch-submitter/pull/73/commits/0c05ec7894ccf196e83ab8e74188d95205e11d24

File: packages/batch-submitter/test/batch-submitter/mockchain-provider.ts [hunk 5]
```diff
@@ -60,25 +61,32 @@ export class MockchainProvider extends providers.JsonRpcProvider {
...
+            queueIndex: 0,
+          },
+        }
+        return info
+      case 'eth_getBlockByNumber':
+        if (params.length === 0) {
+          throw new Error(`Invalid params for ${endpoint}`)
+        }
+        const blockNumber = BigNumber.from((params as any)[0]).toNumber()
+        return this.mockBlocks[blockNumber]
```
Comment (annieke) [inline packages/batch-submitter/test/batch-submitter/mockchain-provider.ts:86]: we might be able to remove this if `injectL2Context` lets us call `getBlockWithTransactions`

**Review by annieke — CHANGES_REQUESTED** [review]
reviewed this with the original PR, everything else looks good!

File: packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts (hunk 2 above)
Comment (gakonst) [reply packages/batch-submitter/src/batch-submitter/state-batch-submitter.ts:208]: Yes that's what the code I linked to above does! The parts of the Optimism Provider made in there are the ones which matter here (injecting extra l2 context)

**Comment by tynes** [comment]
Who owns pushing this over the finishline? @karlfloersch @annieke

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 6]
```diff
@@ -742,70 +706,41 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
-      this.log.error('Current block missing l1BlockNumber', {
-        blockNumber,
-        block,
-      })
+    const batchElement = {
+      ...sequencerFields,
+      stateRoot: block.stateRoot,
+      timestamp: formatNumber(block.timestamp),
+      blockNumber: formatNumber(block.transactions[0].l1BlockNumber),
     }
```
[the remark this answers is not in the document]
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Updated the logic!

[the remark this answers is not in the document]
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Updated the logic!

File: packages/batch-submitter/src/utils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (karlfloersch) [reply packages/batch-submitter/src/utils.ts]: Removed the raw call! Note that we weren't exposing `rawTransaction` before so I added that to the l2 context

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 7]
```diff
@@ -742,70 +706,41 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
-      txData,
-      timestamp: block.timestamp,
-      blockNumber: block.transactions[0].l1BlockNumber,
-    }
+  private async _getBlock(blockNumber: number): Promise<L2Block> {
+    return (await this.l2Provider.send('eth_getBlockByNumber', [
+      toRpcHexString(blockNumber),
+      true,
+    ])) as L2Block
   }
```
[the remark this answers is not in the document]
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Removed the raw call! Note that we weren't exposing rawTransaction before so I added that to the l2 context

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 8]
```diff
@@ -14,6 +14,7 @@ import {
   ctcCoder,
   EthSignTxData,
   txTypePlainText,
+  toRpcHexString,
```
[the remark this answers is not in the document]
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Removed!

**Comment by karlfloersch** [comment]
@tynes I got it

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 3 above)
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Removed all the crappy plaintext things

File: packages/batch-submitter/test/batch-submitter/mockchain-provider.ts (hunk 5 above)
Comment (karlfloersch) [reply packages/batch-submitter/test/batch-submitter/mockchain-provider.ts:86]: Mockchain provider refactor is going to be a beast -- thinking that we should just lump that into the big testing overhaul

File: packages/batch-submitter/src/types.ts [hunk 9]
```diff
@@ -1,23 +1,12 @@
...
 import {
   BlockWithTransactions,
-  Provider,
   TransactionResponse,
 } from '@ethersproject/abstract-provider'
 
-/* Internal Imports */
-import { EIP155TxData, TxType } from '@eth-optimism/core-utils'
-
 export enum QueueOrigin {
```
Comment (tynes) [inline packages/batch-submitter/src/types.ts:7]: Potentially out of scope of this PR but this enum should be moved to `core-utils`

Comment (tynes) [reply packages/batch-submitter/src/types.ts:7]: I notice there is another change to `core-utils` as part of this PR

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 10]
```diff
@@ -742,67 +689,27 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
+      batchElement.isSequencerTx = true
+      batchElement.rawTransaction = block.transactions[0].rawTransaction
     }
+
+    return batchElement
   }
 
   private async _getBlock(blockNumber: number): Promise<L2Block> {
-    const block = (await this.l2Provider.getBlockWithTransactions(
+    return (await this.l2Provider.getBlockWithTransactions(
```
Comment (tynes) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:710]: The `await` here is redundant unless its required for typescript casting but I feel like you could do

```ts
return this.l2Provider.getBlockWithTransactions(blockNumber) as Promise<L2Block>
```

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 11]
```diff
@@ -24,13 +17,7 @@ import {
...
 } from '../transaction-chain-contract'
 
-import {
-  L2Block,
-  BatchElement,
-  Batch,
-  QueueOrigin,
-  queueOriginPlainText,
-} from '..'
+import { L2Block, BatchElement, Batch, QueueOrigin } from '..'
```
Comment (tynes) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:20]: These types could all go in `core-utils`, not opinionated on if its as part of this PR or not

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 12]
```diff
@@ -1,20 +1,13 @@
 /* External Imports */
 import { Promise as bPromise } from 'bluebird'
```
Comment (tynes) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:2]: What do you think of removing `bluebird` as a dependency in the future? Concurrent async calls are easy with `Promise.all`

File: .changeset/sharp-worms-battle.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
+---
+"@eth-optimism/batch-submitter": minor
+---
+
```
Comment (tynes) [inline .changeset/sharp-worms-battle.md:4]: Need changeset for `core-utils`

**Review by tynes — CHANGES_REQUESTED** [review]
The big one is adding the changeset for `core-utils`

File: .changeset/sharp-worms-battle.md (hunk 13 above)
Comment (karlfloersch) [reply .changeset/sharp-worms-battle.md]: Just updated!

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 11 above)
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:20]: Agree they should definitely go into `core-utils`. That said there's a bunch of type & utility cleanup that we could do and so I'm wary of doing a big refactor in the same PR that we're adding this new feature.

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 10 above)
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts]: Btw had to break it into 2 lines in order to pass lint -- the auto-formatting was super buggy using `as Promise<..>` syntax

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 12 above)
Comment (gakonst) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:2]: I had the same thought when I read that code before, here's a post providing good reasoning on "why bluebird" https://blog.kuzzle.io/bluebird-native-async_await-javascript-promises-performances-2020

File: .changeset/witty-buses-worry.md [hunk 14]
```diff
@@ -0,0 +1,7 @@
+---
+"@eth-optimism/batch-submitter": minor
+"@eth-optimism/core-utils": patch
+---
+
+- Use raw transaction in batch submitter -- incompatible with L2Geth v0.1.2.1
```
Comment (gakonst) [inline .changeset/witty-buses-worry.md:6]: What is geth 0.1.2.1? Is it deployed / being used anywhere right now?

**Review by gakonst — APPROVED** [review]
Code changes LGTM - assuming they are not breaking Kovan or Mainnet.

File: .changeset/witty-buses-worry.md (hunk 14 above)
Comment (karlfloersch) [reply .changeset/witty-buses-worry.md:6]: Ah it is the version of Geth that we were running before our last regenesis ~2/3 weeks ago. This should not be breaking Kovan or Mainnet now that we've got `rawTransaction` in mainnet

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 11 above)
Comment (annieke) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:20]: ^happy to help with the utility cleanup!

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts [hunk 15]
```diff
@@ -742,67 +689,26 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
...
-        sequencerTxType: undefined,
-        txData: undefined,
-        timestamp: block.timestamp,
-        blockNumber: block.transactions[0].l1BlockNumber,
-      }
+    const batchElement = {
+      stateRoot: block.stateRoot,
+      timestamp: block.timestamp,
+      blockNumber: block.transactions[0].l1BlockNumber,
+      isSequencerTx: undefined,
```
Comment (annieke) [inline packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:697]: should this be false?
```suggestion
      isSequencerTx: false,
```

Comment (annieke) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:697]: and is `txData` deprecated?

File: packages/batch-submitter/test/batch-submitter/mockchain-provider.ts (hunk 5 above)
Comment (annieke) [reply packages/batch-submitter/test/batch-submitter/mockchain-provider.ts:86]: noted!

**Review by annieke — APPROVED** [review]
looks good! yay for deleted code, only concern is whether `isSequencerTx` is set correctly

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 15 above)
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:697]: Yep! `txData` and `txType` are both deprecated! 

Also, yes I think it's better for this to be false! Good catch -- updating

File: packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts (hunk 11 above)
Comment (karlfloersch) [reply packages/batch-submitter/src/batch-submitter/tx-batch-submitter.ts:20]: 🙏🙏🙏

**Review by tynes — APPROVED** [review]
Looks good to me!
